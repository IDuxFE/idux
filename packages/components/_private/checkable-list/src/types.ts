/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
