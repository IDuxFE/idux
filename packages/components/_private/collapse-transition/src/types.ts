/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
