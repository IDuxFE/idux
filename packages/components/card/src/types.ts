import type { ComponentSize } from '@idux/components/core/types'
export interface CardProps {
  // please add readonly for every prop
  readonly title?: string
  readonly extra?: string
  readonly hoverable: boolean
  readonly bordered: boolean
  readonly loading: boolean
  readonly size?: ComponentSize
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IxCardComponent extends CardProps {}
