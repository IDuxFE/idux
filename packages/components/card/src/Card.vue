<template>
  <div class="ix-card" :class="classes">
    <div v-if="isShowLoading" class="ix-card-loading-block"></div>
    <div v-else class="ix-card-wraper">
      <div class="ix-card__head">
        <div v-if="isShowTitle" class="ix-card__head__title">
          <slot name="title">
            <a>{{ title }}</a>
          </slot>
        </div>
        <div v-if="isShowExtra" class="ix-card__head__extra">
          <slot name="extra">
            <a>{{ extra }}</a>
          </slot>
        </div>
      </div>
      <div class="ix-card__body">
        <slot></slot>
      </div>
      <div v-if="isShowFooter" class="ix-card__footer">
        <slot name="footer"></slot>
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import type { PropType } from 'vue'
import type { ComponentSize } from '@idux/components/core/types'
import type { CardConfig } from '@idux/components/core/config/types'

import { useGlobalConfig } from '@idux/components/core/config'
import { computed, defineComponent } from 'vue'
import { CardProps } from './types'

export default defineComponent({
  name: 'IxCard',
  props: {
    title: { type: String, default: undefined },
    extra: { type: String, default: undefined },
    hoverable: Boolean,
    borderless: Boolean,
    loading: Boolean,
    size: { type: String as PropType<ComponentSize>, default: undefined },
  },
  setup(props: CardProps, { slots }) {
    // init
    const cardConfig = useGlobalConfig('card')
    const classes = useClasses(props, cardConfig)
    const isShowLoading = computed(() => (props.loading !== undefined ? props.loading : cardConfig.loading))
    const isShowTitle = computed(() => typeof props.title === 'string' || !!slots.title)
    const isShowExtra = computed(() => typeof props.extra === 'string' || !!slots.extra)
    const isShowFooter = computed(() => !!slots.extra)
    return { classes, isShowTitle, isShowExtra, isShowFooter, isShowLoading }
  },
})

const useClasses = (props: CardProps, cardConfig: CardConfig) => {
  return computed(() => {
    const hoverable = props.hoverable !== undefined ? props.hoverable : cardConfig.hoverable
    const borderless = props.borderless !== undefined ? props.borderless : cardConfig.borderless
    const size = props.size !== undefined ? props.size : cardConfig.size
    return [
      size !== 'medium' ? `ix-card-${size}` : '',
      {
        'ix-card-hover': hoverable,
        'ix-card-border': borderless,
      },
    ]
  })
}
</script>
