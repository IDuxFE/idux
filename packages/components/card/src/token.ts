import type { ComputedRef, InjectionKey } from 'vue'

export interface CardContext {
  hoverable: ComputedRef<boolean>
  registerGrid: () => void
  unregisterGrid: () => void
}

export const cardToken: InjectionKey<CardContext> = Symbol('cardToken')
