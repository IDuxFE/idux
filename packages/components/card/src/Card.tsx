/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { CardProps } from './types'
import type { ComputedRef, Slots, VNode, VNodeTypes } from 'vue'

import { computed, defineComponent, isVNode, normalizeClass, provide } from 'vue'

import { isString } from 'lodash-es'

import { ɵHeader } from '@idux/components/_private/header'
import { IxButton } from '@idux/components/button'
import { useGlobalConfig } from '@idux/components/config'
import { IxCol, IxRow } from '@idux/components/grid'

import { cardToken } from './token'
import { cardProps } from './types'

export default defineComponent({
  name: 'IxCard',
  props: cardProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-card`)
    const config = useGlobalConfig('card')
    const hoverable = computed(() => props.hoverable ?? config.hoverable)

    provide(cardToken, { hoverable })

    const children = computed(() => slots.default?.() || [])
    const hasGrid = computed(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return children.value.some(node => node.type && (node.type as any).name === 'IxCardGrid')
    })

    const classes = computed(() => {
      const { borderless = config.borderless, loading, size = config.size } = props
      const hasGridValue = hasGrid.value
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-borderless`]: borderless,
        [`${prefixCls}-hoverable`]: !hasGridValue && hoverable.value,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-has-grid`]: hasGridValue,
        [`${prefixCls}-${size}`]: true,
      })
    })

    return () => {
      const prefixCls = mergedPrefixCls.value
      return (
        <div class={classes.value}>
          {renderCover(props, slots, prefixCls)}
          <ɵHeader v-slots={slots} header={props.header} />
          {renderBody(props, children, hasGrid, prefixCls)}
          {renderFooter(props, slots, prefixCls)}
        </div>
      )
    }
  },
})

const renderCover = (props: CardProps, slots: Slots, prefixCls: string) => {
  let coverNode: VNodeTypes | undefined
  if (slots.cover) {
    coverNode = slots.cover()
  } else if (props.cover) {
    const { cover } = props
    const imgProps = isString(cover) ? { src: cover } : cover
    coverNode = <img {...imgProps} />
  }
  return coverNode ? <div class={`${prefixCls}-cover`}>{coverNode}</div> : undefined
}

const renderBody = (
  props: CardProps,
  children: ComputedRef<VNode[]>,
  hasGrid: ComputedRef<boolean>,
  prefixCls: string,
) => {
  let bodyNode: VNodeTypes | undefined
  if (props.loading) {
    bodyNode = renderLoading(prefixCls)
  } else if (children.value.length) {
    bodyNode = hasGrid.value ? <IxRow>{children.value}</IxRow> : children.value
  }
  return bodyNode ? <div class={`${prefixCls}-body`}>{bodyNode}</div> : undefined
}

const loadingSpans = [[22], [8, 15], [6, 18], [13, 9], [4, 3, 16], [8, 6, 8]]

const renderLoading = (prefixCls: string) => {
  const loadingChild = loadingSpans.map(spans => {
    const cols = spans.map(span => (
      <IxCol span={span} class={`${prefixCls}-loading-col`}>
        <div class={`${prefixCls}-loading-block`}></div>
      </IxCol>
    ))
    return <IxRow>{cols}</IxRow>
  })
  return <div class={`${prefixCls}-loading`}>{loadingChild}</div>
}

const renderFooter = (props: CardProps, slots: Slots, prefixCls: string) => {
  if (!slots.footer && !props.footer) {
    return undefined
  }

  let footerNode: VNodeTypes
  if (slots.footer) {
    footerNode = slots.footer()
  } else {
    footerNode = props.footer!.map(item => {
      let itemChild: VNodeTypes | null
      if (isVNode(item)) {
        itemChild = item
      } else {
        const { text, ...rest } = item
        itemChild = <IxButton {...rest}>{text}</IxButton>
      }
      return itemChild ? <li>{itemChild}</li> : null
    })
  }

  return <ul class={`${prefixCls}-footer`}>{footerNode}</ul>
}
