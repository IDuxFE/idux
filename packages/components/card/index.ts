/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { CardComponent, CardGridComponent } from './src/types'

import Card from './src/Card'
import CardGrid from './src/CardGrid'

const IxCard = Card as unknown as CardComponent
const IxCardGrid = CardGrid as unknown as CardGridComponent

export { IxCard, IxCardGrid }

export type {
  CardInstance,
  CardPublicProps as CardProps,
  CardGridInstance,
  CardGridPublicProps as CardGridProps,
  CardSize,
  CardCover,
  CardButtonProps,
} from './src/types'
