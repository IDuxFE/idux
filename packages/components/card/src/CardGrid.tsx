/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, normalizeClass } from 'vue'

import { useGlobalConfig } from '@idux/components/config'
import { IxCol } from '@idux/components/grid'

import { cardToken } from './token'
import { cardGridProps } from './types'

export default defineComponent({
  name: 'IxCardGrid',
  props: cardGridProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-card-grid`)
    const { hoverable } = inject(cardToken)!

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-hoverable`]: props.hoverable ?? hoverable.value,
      })
    })

    return () => <IxCol class={classes.value}>{slots.default?.()}</IxCol>
  },
})
