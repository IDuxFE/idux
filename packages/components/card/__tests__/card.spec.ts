import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import { IxButton } from '@idux/components/button'
import { IxHeader } from '@idux/components/header'

import Card from '../src/Card'
import { CardProps } from '../src/types'

const logoBase64 =
  'data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB3aWR0aD0iMTI4cHgiIGhlaWdodD0iMTI4cHgiIHZpZXdCb3g9IjAgMCAxMjggMTI4IiB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiPgogICAgPCEtLSBHZW5lcmF0b3I6IFNrZXRjaCA1OCAoODQ2NjMpIC0gaHR0cHM6Ly9za2V0Y2guY29tIC0tPgogICAgPHRpdGxlPjEyODwvdGl0bGU+CiAgICA8ZGVzYz5DcmVhdGVkIHdpdGggU2tldGNoLjwvZGVzYz4KICAgIDxkZWZzPgogICAgICAgIDxsaW5lYXJHcmFkaWVudCB4MT0iNTAlIiB5MT0iMCUiIHgyPSI1MCUiIHkyPSI5OS42Mjc4NTA1JSIgaWQ9ImxpbmVhckdyYWRpZW50LTEiPgogICAgICAgICAgICA8c3RvcCBzdG9wLWNvbG9yPSIjMDBBQ0ZGIiBvZmZzZXQ9IjAlIj48L3N0b3A+CiAgICAgICAgICAgIDxzdG9wIHN0b3AtY29sb3I9IiMzMzY2RkYiIG9mZnNldD0iMTAwJSI+PC9zdG9wPgogICAgICAgIDwvbGluZWFyR3JhZGllbnQ+CiAgICAgICAgPGxpbmVhckdyYWRpZW50IHgxPSI1MCUiIHkxPSIwJSIgeDI9IjUwJSIgeTI9IjEwMCUiIGlkPSJsaW5lYXJHcmFkaWVudC0yIj4KICAgICAgICAgICAgPHN0b3Agc3RvcC1jb2xvcj0iIzFEQjgzRiIgb2Zmc2V0PSIwJSI+PC9zdG9wPgogICAgICAgICAgICA8c3RvcCBzdG9wLWNvbG9yPSIjNzJEMTNEIiBvZmZzZXQ9IjEwMCUiPjwvc3RvcD4KICAgICAgICA8L2xpbmVhckdyYWRpZW50PgogICAgPC9kZWZzPgogICAgPGcgaWQ9IjEyOCIgc3Ryb2tlPSJub25lIiBzdHJva2Utd2lkdGg9IjEiIGZpbGw9Im5vbmUiIGZpbGwtcnVsZT0iZXZlbm9kZCI+CiAgICAgICAgPHBhdGggZD0iTTQ0LjQ2NzMxMjYsNjIuMjMyMTcwNSBMNzAuMTI5ODAxNSw4My43NjU1NTU0IEM2MC4xODk3NTEyLDk1LjYxMTY0NiA0Mi41Mjg1OTQ2LDk3LjE1Njc5NyAzMC42ODI1MDQsODcuMjE2NzQ2NyBMMjguMjQ4MTQ5Miw4NS4xNzM4OTc1IEMyMS4zMTIxMTA4LDc5LjMwODkyOTQgMTQuMzM1NDQzOSw3My40NTQ4NzE4IDcuMzE4MTQ4NDYsNjcuNjExNzI0NiBMOS4yNDY1MTEyOSw2NS4zMTM1OTEzIEMxMy44NDgzODY0LDU5LjgyOTI5MDEgMjAuMzAzNzUwNyw1Ni44MTc3Nzk2IDI2LjkyNzcxNzYsNTYuNDI2NDY4MyBMMjguMzQ5MTA1Nyw1Ni4zODI2OTk1IEMzNC4wNDAwNjg2LDU2LjM2ODg1MDQgMzkuNzY2NDgzLDU4LjI4NzcwNjEgNDQuNDY3MzEyNiw2Mi4yMzIxNzA1IFoiIGlkPSLot6/lvoQiIGZpbGw9IiMyMDRFRDkiPjwvcGF0aD4KICAgICAgICA8cGF0aCBkPSJNOTUuNDM4MDk4NCw0NS4xMzQ1MDkgTDk3LjMzMTY2NTYsNDUuMjI3OTkzMiBMOTguNzAxMDU1OSw0NS4zNjI3MDcyIEwxMDAuNTE2MTI5LDQ1LjYyOTI5MTYgTDEwMS44OTEyODcsNDUuODk5NDE2NCBMMTAzLjQzNjU4NSw0Ni4yNzUwNzE5IEwxMDUuMTY5NDEsNDYuNzkwNDA1MSBMMTA2LjU4NzY2NSw0Ny4yODk1OTM3IEwxMDguMjEyNTY1LDQ3Ljk1MjAwNzQgTDEwOC45MzE3Niw0OC4yNzc2NDAxIEwxMDguOTMxNzYsNDguMjc3NjQwMSBMMTEwLjM1Mzg5OCw0OC45ODM2MzEgTDExMS44MDUxMjgsNDkuNzk0NjA0NSBMMTEzLjQ0MzE5OSw1MC44MzAxOTgxIEwxMTQuNjUzNDk2LDUxLjY4NTY5MTIgTDExNS44MjE1MjEsNTIuNTkyMTI3NSBMMTE3LjE4ODUzNyw1My43NjU3NjYgTDExOC40NzMyOTksNTQuOTk1MDg2NCBMMTE5LjUzMzEzLDU2LjExNTU4MDYgTDEyMC4wODg2Miw1Ni43NDY0NjgzIEwxMjAuMDg4NjIsNTYuNzQ2NDY4MyBMMTIxLjIwODAwMyw1OC4xMjIyMDA1IEwxMjIuMTMwNDY5LDU5LjM3NzQ1NiBMNzIuMjk3NDUzMyw5NC4yNzA5MDkyIEM2Ny41NzI1ODc4LDk3LjU3OTI5NTcgNjIuMTMzNTU1NCw5OS4zMTAyNzggNTYuNjY4NDkzMSw5OS40OTg0MDA0IEw1NS4wMjg4ODE3LDk5LjUwODU4NTggQzQ4LjQ3MjgxNzUsOTkuMzY0NDYwNyA0MS45NzgzNjIyLDk3LjAwMzU5MTYgMzYuNzM0NjE2NSw5Mi40ODU2NzA5IEwzMi43NjgwODE2LDg5LjAzMzg2MiBMMzIuNzY4MDgxNiw4OS4wMzM4NjIgTDI3LjYzNjUwMzksODQuNjQ2ODk0IEw3NS42Mzg0ODQ5LDUxLjE0MzQ2MzQgQzc3LjQ1OTY4NTEsNDkuODcyMzM5NyA3OS4zNjM0NjM3LDQ4LjgwNzg1NjQgODEuMzIxOTc2Miw0Ny45NDUwODIgTDgzLjI5NzM3NzksNDcuMTQ5NDc5OSBDODQuMzIzMzMxMSw0Ni43NzQwNzgyIDg1LjM2MTE5MzEsNDYuNDUxNzgxNSA4Ni40MDcyNTc2LDQ2LjE4MTg4MTUgTDg3Ljk4MjA0MzEsNDUuODE2MjQwMyBMODkuNjQ0MTg2Niw0NS41MTY1NyBMOTEuMzg1OTMwMyw0NS4yOTQ4NTUxIEw5My4wNTA0MjI4LDQ1LjE2OTM0MjMgQzkzLjg0NjY4MzIsNDUuMTI5Mjc1NyA5NC42NDMxMDY0LDQ1LjExNzc1ODggOTUuNDM4MDk4NCw0NS4xMzQ1MDkgWiIgaWQ9Iui3r+W+hCIgZmlsbD0idXJsKCNsaW5lYXJHcmFkaWVudC0xKSI+PC9wYXRoPgogICAgICAgIDxwYXRoIGQ9Ik01MS4zODg4NTIsMzIuNDY5NDc1OSBMNjMuNjQ1NTYzMSw0Mi43NTQwNzc2IEM1OC42NzU1MzgsNDguNjc3MTIyOSA0OS44NDQ5NTk3LDQ5LjQ0OTY5ODQgNDMuOTIxOTE0NCw0NC40Nzk2NzMzIEwzMi44MTQyNywzNS4xNTkyNTMgTDMzLjc3ODQ1MTQsMzQuMDEwMTg2MyBDMzguMjE1OTczOCwyOC43MjE3NTMgNDYuMTAwNDE4NywyOC4wMzE5NTM0IDUxLjM4ODg1MiwzMi40Njk0NzU5IFoiIGlkPSLot6/lvoQiIGZpbGw9IiMwMzc4MkEiPjwvcGF0aD4KICAgICAgICA8cGF0aCBkPSJNNzIuMDY0NTA2MiwyNi4xMjg0NDA2IEw3My4yNjYzMDc5LDI2LjIyMjA0MTMgTDc0LjM3NzU0MTYsMjYuMzk3MDkgTDc1LjU3NjI3OTIsMjYuNjg0OTIyNCBMNzYuNDc4NTYyMywyNi45NzMxMDM2IEw3Ny41MTY0NTg5LDI3LjM4NjMzNDcgTDc3LjUxNjQ1ODksMjcuMzg2MzM0NyBMNzguMjAwNjU1MywyNy43MTA0NDczIEw3OS4zMDc1MTExLDI4LjMzMTQwODYgTDc5LjcwNDIsMjguNTg1OTgwNSBMODAuMjkxNTU3OCwyOC45OTc0NTY2IEw4MC44MDkzMjg3LDI5LjM5NzU4MjggTDgxLjIxODA3NjgsMjkuNzQwODE0MyBMODEuNzcyNzI0MywzMC4yNDk2ODA3IEw4Mi4yMTE2Mzk2LDMwLjY5MTk2MzIgTDgyLjIxMTYzOTYsMzAuNjkxOTYzMiBMODIuNjM0NzgxOCwzMS4xNTYxOTk0IEw4My4wOTIyMjcyLDMxLjcwNjIzMjEgTDgzLjYxOTAyNjMsMzIuNDEzNDQ1NSBMNjMuMTYzMzEwOCw0Ni43MzY2OTE3IEM2MC45ODg3ODEsNDguMjU5MzEzOSA1OC41MTE3ODE3LDQ5LjExMzY3NjggNTYuMDAwNjA2Myw0OS4zMTMzMzE2IEw1NC43NDM0MTI4LDQ5LjM1ODY2MjggQzUxLjM4OTkyNjQsNDkuMzM0Mzg5NyA0OC4wNTUxNzcxLDQ4LjE1MDIwNjEgNDUuMzc1MTE5NSw0NS44MzgyMzM0IEw0Mi4zNzc0OTk1LDQzLjI1MjMxMyBMNjMuMDcwNjgxNiwyOC43ODMyMDI2IEM2My42NzAxMjExLDI4LjM2NDA2MTggNjQuMjg5ODcxMywyNy45OTU2NDMzIDY0LjkyNDgzNTYsMjcuNjc3MDQ2OCBMNjUuOTUxNDUxMywyNy4yMTAyOTkxIEw2NS45NTE0NTEzLDI3LjIxMDI5OTEgTDY2LjU3MDYwOTYsMjYuOTczMjE1NyBMNjcuNzY3NzYyNywyNi42MDI1NDE2IEw2Ny43Njc3NjI3LDI2LjYwMjU0MTYgTDY4Ljk4MjU5NjQsMjYuMzM3MjA5MiBMNjkuNzk2ODcyMSwyNi4yMTg1ODE3IEw3MC4yMTY0MDg1LDI2LjE3NTQ3ODUgQzcwLjgzMjMzODgsMjYuMTIxMDkyMyA3MS40NDk2MzQ3LDI2LjEwNTYzNjEgNzIuMDY0NTA2MiwyNi4xMjg0NDA2IFoiIGlkPSLot6/lvoQiIGZpbGw9InVybCgjbGluZWFyR3JhZGllbnQtMikiPjwvcGF0aD4KICAgIDwvZz4KPC9zdmc+'

describe('Card', () => {
  const CardMount = (options?: MountingOptions<Partial<CardProps>>) => mount(Card, { ...options })

  renderWork<CardProps>(Card, { slots: { default: () => h('p', 'Some Content') } })

  test('borderless work', async () => {
    const wrapper = CardMount({ props: { borderless: true } })
    expect(wrapper.classes()).toContain('ix-card-borderless')

    await wrapper.setProps({ borderless: false })
    expect(wrapper.classes()).not.toContain('ix-card-borderless')
  })

  test('cover work', async () => {
    const wrapper = CardMount({ props: { cover: logoBase64 } })
    expect(wrapper.find('.ix-card-cover').exists()).toBe(true)
    expect(wrapper.find('img').element.src).toBe(logoBase64)

    const alt = 'alt'
    const srcset = 'srcset.svg'

    await wrapper.setProps({ cover: { src: logoBase64, alt, srcset } })

    expect(wrapper.find('.ix-card-cover').exists()).toBe(true)
    expect(wrapper.find('img').element.src).toBe(logoBase64)
    expect(wrapper.find('img').element.alt).toBe(alt)
    expect(wrapper.find('img').element.srcset).toBe(srcset)
  })

  test('cover slot work', async () => {
    const overSlot = 'cover slot'
    const wrapper = CardMount({
      props: { cover: logoBase64 },
      slots: { cover: () => overSlot },
    })
    expect(wrapper.find('.ix-card-cover').text()).toBe(overSlot)
    expect(wrapper.find('img').exists()).toBe(false)
  })

  test('header work', async () => {
    const header = 'title'
    const wrapper = CardMount({ props: { header } })
    expect(wrapper.find('.ix-header').text()).toBe(header)

    await wrapper.setProps({ header: { title: header, suffix: 'up' } })

    expect(wrapper.find('.ix-header').text()).toBe(header)
    expect(wrapper.find('.ix-icon-up').exists()).toBe(true)
  })

  test('header slot work', async () => {
    const header = 'title'
    const wrapper = CardMount({
      props: { header },
      slots: { header: () => h(IxHeader, { title: 'header slot' }) },
    })

    expect(wrapper.find('.ix-header').text()).toBe('header slot')
  })

  test('hoverable work', async () => {
    const wrapper = CardMount({ props: { hoverable: true } })
    expect(wrapper.classes()).toContain('ix-card-hoverable')

    await wrapper.setProps({ hoverable: false })
    expect(wrapper.classes()).not.toContain('ix-card-hoverable')
  })

  test('shadow work', async () => {
    const wrapper = CardMount()
    expect(wrapper.classes()).toContain('ix-card-has-shadow')

    await wrapper.setProps({ shadow: false })
    expect(wrapper.classes()).not.toContain('ix-card-has-shadow')
  })

  test('disabled work', async () => {
    const wrapper = CardMount({ props: { disabled: true } })
    expect(wrapper.classes()).toContain('ix-card-disabled')

    await wrapper.setProps({ disabled: false })
    expect(wrapper.classes()).not.toContain('ix-card-disabled')
  })

  test('selectable work', async () => {
    const wrapper = CardMount({ props: { selectable: true } })
    expect(wrapper.classes()).toContain('ix-card-selectable')

    await wrapper.setProps({ selectable: false })
    expect(wrapper.classes()).not.toContain('ix-card-selectable')
  })

  test('v-model:selected work', async () => {
    const onUpdateSelected = vi.fn()
    const wrapper = CardMount({ props: { selectable: true, selected: true, 'onUpdate:selected': onUpdateSelected } })
    expect(wrapper.classes()).toContain('ix-card-selected')

    await wrapper.setProps({ selected: false })
    expect(wrapper.classes()).not.toContain('ix-card-selected')

    await wrapper.trigger('click')
    expect(onUpdateSelected).toBeCalledWith(true)
  })

  test('onChange work', async () => {
    const handleSelected = vi.fn()
    const wrapper = CardMount({
      props: {
        onSelectedChange: handleSelected,
        selectable: true,
        selected: false,
      },
    })

    await wrapper.trigger('click')
    expect(handleSelected).toBeCalledWith(true)
  })

  test('loading work', async () => {
    const wrapper = CardMount({ props: { loading: true } })
    expect(wrapper.find('.ix-card-loading').exists()).toBe(true)

    await wrapper.setProps({ loading: false })
    expect(wrapper.find('.ix-card-loading').exists()).not.toBe(true)
  })

  test('size work', async () => {
    const wrapper = CardMount({ props: { size: 'sm' } })

    expect(wrapper.classes()).toContain('ix-card-sm')

    await wrapper.setProps({ size: 'lg' })

    expect(wrapper.classes()).toContain('ix-card-lg')

    await wrapper.setProps({ size: 'md' })

    expect(wrapper.classes()).toContain('ix-card-md')
  })

  test('footer work', async () => {
    const onClick0 = vi.fn()
    const onClick1 = vi.fn()
    const footer = [
      { icon: 'up', onClick: onClick0 },
      { icon: 'down', onClick: onClick1 },
    ]
    const wrapper = CardMount({ props: { footer } })

    const buttons = wrapper.find('.ix-card-footer').findAll('.ix-button')

    expect(buttons.length).toBe(2)

    await buttons[0].trigger('click')

    expect(onClick0).toBeCalled()

    await buttons[1].trigger('click')

    expect(onClick1).toBeCalled()
  })

  test('footer slot work', async () => {
    const footer = [{ icon: 'up' }, { icon: 'down' }]
    const wrapper = CardMount({
      props: { footer },
      slots: { footer: () => [h(IxButton, { icon: 'left' }), h(IxButton, { icon: 'right' })] },
    })

    expect(wrapper.find('.ix-card-footer').find('.ix-icon-left').exists()).toBe(true)
    expect(wrapper.find('.ix-card-footer').find('.ix-icon-right').exists()).toBe(true)
    expect(wrapper.find('.ix-card-footer').find('.ix-icon-up').exists()).toBe(false)
    expect(wrapper.find('.ix-card-footer').find('.ix-icon-down').exists()).toBe(false)
  })
})
