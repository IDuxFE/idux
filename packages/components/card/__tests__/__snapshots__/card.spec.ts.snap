// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue render work 1`] = `"<div class=\\"ix-card\\"><div class=\\"ix-card-wraper\\"><div class=\\"ix-card__head\\"><!--v-if--><!--v-if--></div><div class=\\"ix-card__body\\"></div><!--v-if--></div></div>"`;
